File: 07.Week 6 - Python/Credit.py
import sys


def main():
    # Prompt the user for a credit card number (accept as string to preserve leading zeros)
    while True:
        try:
            number = int(input("Number: "))
            if number > 0:
                break
        except ValueError:
            pass

    # Convert to string for easy digit manipulation
    digits = str(number)
    length = len(digits)

    # ---------- Luhn's Algorithm ----------
    # Starting from the second-to-last digit, double every other digit
    total = 0
    for i, ch in enumerate(reversed(digits)):
        d = int(ch)
        if i % 2 == 1:          # every second digit from the right (0-indexed)
            d *= 2
            if d > 9:
                d -= 9          # equivalent to summing the two resulting digits
        total += d

    # If the total does not end in 0, the number is invalid
    if total % 10 != 0:
        print("INVALID")
        return

    # ---------- Identify the card brand ----------
    # AMEX:        starts with 34 or 37, length 15
    # MASTERCARD:  starts with 51-55,    length 16
    # VISA:        starts with 4,        length 13 or 16

    if length == 15 and digits[:2] in ("34", "37"):
        print("AMEX")
    elif length == 16 and 51 <= int(digits[:2]) <= 55:
        print("MASTERCARD")
    elif digits[0] == "4" and length in (13, 16):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: 02.Week 1 - C/credit.c
#include <stdio.h>
#include <math.h>
#include <stdbool.h>

int main(void)
{
    long long number;
    printf("Number: ");
    scanf("%lld", &number);

    // Count digits and check validity
    long long temp = number;
    int digits = 0;
    while (temp > 0)
    {
        digits++;
        temp /= 10;
    }

    // Luhn's Algorithm
    long long n = number;
    int sum = 0;
    bool alternate = false;

    while (n > 0)
    {
        int digit = n % 10;

        if (alternate)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }

        sum += digit;
        alternate = !alternate;
        n /= 10;
    }

    // Check starting digits and length
    long long first_two = number;
    while (first_two > 99)
    {
        first_two /= 10;
    }

    long long first_four = number;
    while (first_four > 9999)
    {
        first_four /= 10;
    }

    if (sum % 10 == 0)
    {
        if ((first_two == 34 || first_two == 37) && digits == 15)
        {
            printf("AMEX\n");
        }
        else if (first_four >= 5100 && first_four <= 5500 && digits == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((first_two / 10) == 4 && (digits == 13 || digits == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
